File: store.py
from typing import Optional, List

class Store:

    def get(self, key):
        pass

    def put(self, key, value):
        pass

    def delete(self, key):
        pass
    
class KeyValueStore(Store):

    def __init__(self, base_store: Optional[Store]):
        self.base_store = base_store
        self.deletions = set()
        self.data = {}

    def contains(self, key):
        if key not in self.data:
            if self.base_store and self.base_store.contains(key):
                return key not in self.deletions
        return key in self.data

    def reset(self):
        self.data = {}
        self.deletions = {}

    def get(self, key):
        if key not in self.data:
            if self.base_store and key not in self.deletions:
                return self.base_store.get(key)
        return self.data.get(key, None)

    def put(self, key, value):
        if self.base_store:
            self.deletions.discard(key)
        self.data[key] = value

    def delete(self, key):
        if self.base_store:
            self.deletions.add(key)
        self.data.pop(key, None)


class MemoryStore:

    def __init__(self):
        self.layers : List[KeyValueStore] = []
        self._curr_index = 0
        self.layers.append(KeyValueStore(None))

    def begin(self):
        self._curr_index += 1
        layer = KeyValueStore(self.layers[self._curr_index - 1])
        self.layers.append(layer)

    def _merge_layer(self, base_layer, transaction_layer):
        for key in transaction_layer.deletions:
            base_layer.delete(key)
        for key, value in transaction_layer.data.items():
            base_layer.put(key, value)
        transaction_layer.reset()

    def commit(self):
        if self._curr_index == 0:
            raise Exception("No active Transaction")
        transaction_layer = self.layers.pop()
        self._curr_index -= 1
        
        self._merge_layer(self.layers[self._curr_index], transaction_layer)


    def rollback(self):
        if self._curr_index == 0:
            raise Exception("No active Transaction")
        _ = self.layers.pop()
        self._curr_index -= 1

    def get(self, key):
        return self.layers[self._curr_index].get(key)

    def put(self, key, value):
        self.layers[self._curr_index].put(key, value)

    def delete(self, key):
        self.layers[self._curr_index].delete(key)


if __name__ == "__main__":

    # Basic Test
    store = MemoryStore()
    store.put("A", 1)
    assert store.get("A") == 1
    assert store.get("B") == None
    store.delete("A")
    assert store.get("A") == None

    # Transaction tests
    store = MemoryStore()
    store.put("A", 1)
    store.begin()
    store.delete("A")

    store.begin()
    store.put("B", 2)
    store.commit()
    store.put("A", 3)

    assert store.get("A") == 3

    store.delete("A")
    store.commit()
    assert store.get("A") == None
    assert store.get("B") == 2
    

File: Store.cpp
#include<iostream>
#include<unordered_map>
#include<set>
#include<vector>
#include<string>

template<typename KeyT, typename ValueT>
class Store {
    public:
        virtual const ValueT * const get(const KeyT & key) const;

        virtual void put(KeyT key, ValueT value);

        virtual bool remove(const KeyT & key);
};

template<typename KeyT, typename ValueT>
class TransactionStore : public Store<KeyT, ValueT>
{
    public:
        virtual void begin();

        virtual void commit();

        virtual void rollback();
};



template<typename KeyT, typename ValueT>
class KeyValueStore : public Store<KeyT, ValueT> {
    private:
        std::unordered_map<KeyT, ValueT> data;
        std::set<KeyT> deletions;
        Store<KeyT, ValueT> * baseStore = nullptr;
    public:

        KeyValueStore(Store<KeyT, ValueT> * baseStore = nullptr) {
            this->baseStore = baseStore;
        }

        void saveToBaseStore() {
            if(this->baseStore == nullptr) {
                return;
            }

            for(auto it = this->deletions.begin(); it != this->deletions.end(); ++it) {
                this->baseStore->remove(*it);
            }

            for(auto it = this->data.begin(); it != this->data.end(); ++it) {
                this->baseStore->put(it->first, it->second);
            }

            this->data.clear();
            this->deletions.clear();
        }

        const ValueT * const get(const KeyT & key) const {
            if( this->deletions.find(key) != this->deletions.end() ) {
                return nullptr;
            }
            auto it = this->data.find(key);
            if(  it != this->data.end() ) {
                return & it->second;
            }
            if(this->baseStore != nullptr) {
                return this->baseStore->get(key);
            }
            return nullptr;
        }

        void put(KeyT key, ValueT value) {
            if(this->baseStore != nullptr) {
                this->deletions.erase(key);
            }
            this->data.insert_or_assign(key, value);
        }

        bool remove(KeyT key) {
            if(this->baseStore != nullptr) {
                this->deletions.insert(key);
            }
            return this->data.erase(key) == 1;
        }
};


template<typename KeyT, typename ValueT>
class MemoryStore : public TransactionStore<KeyT, ValueT> {
    private:
        std::vector<KeyValueStore<KeyT, ValueT>*> dataLayers;
        int currIndex = 0;
    public:
        MemoryStore() {
            KeyValueStore<KeyT, ValueT> * layer = new KeyValueStore<KeyT, ValueT>();
            dataLayers.push_back(layer);
        }

        void begin() {
            KeyValueStore<KeyT, ValueT> * baseLayer = this->dataLayers[this->currIndex];
            KeyValueStore<KeyT, ValueT> * layer = new KeyValueStore<KeyT, ValueT>(baseLayer);
            this->currIndex += 1;
        }

        void commit() {
            if(this->currIndex == 0) {
                throw std::runtime_error("Cannot commit: no active transaction");
            }
            KeyValueStore<KeyT, ValueT> * transactionLayer = this->dataLayers[this->currIndex];
            transactionLayer->saveToBaseStore();
            this->dataLayers.pop_back();
            delete transactionLayer;
            this->currIndex -= 1;
        }

        void rollback() {
            if(this->currIndex == 0) {
                throw std::runtime_error("Cannot rollback: no active transaction");
            }
            KeyValueStore<KeyT, ValueT> * transactionLayer = this->dataLayers[this->currIndex];
            this->dataLayers.pop_back();
            delete transactionLayer;
            this->currIndex -= 1;
        }

        const ValueT * const get(const KeyT & key) const {
            return this->dataLayers[this->currIndex]->get(key);
        }

        void put(KeyT key, ValueT value) {
            this->dataLayers[this->currIndex]->put(key, value);
        }

        bool remove(const KeyT & key) {
            return this->dataLayers[this->currIndex]->remove(key);
        }

    
};

int main() {
    MemoryStore<std::string, int> store;

    return 0;
}